### PR TITLE
Fix chi2 calculation for annulus measurements

### PR DESCRIPTION
**Pull request by shimasnd**
Update of `track_state.hpp` to get annulus measurements properly in track_state, otherwise chi2 is always zero for annulus measurements.

### CODE REVIEW

File: core/include/traccc/edm/track_state.hpp
Comment (krasznaa) [inline core/include/traccc/edm/track_state.hpp:80]: You should add a line of documentation that explains what this check is actually doing. Since I don't know this code, this line looks like checking some "magic number". :frowning:

**Review by krasznaa — CHANGES_REQUESTED** [review]
@beomki-yeo will have to say whether the change is good or not. Formally, it looks fine. But some comments are desperately needed in the code for the lay reader.

File: core/include/traccc/edm/track_state.hpp
Comment (beomki-yeo) [reply core/include/traccc/edm/track_state.hpp:80]: I had to rewind the logic from scratch...
What I think now is that we should not touch `track_state` at all

@shimasnd Have you checked the matrix elements of the projection matrix? https://github.com/acts-project/traccc/blob/fd145d8fa3b5dac40fa88e449d1d235f0dff6b85/core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp#L74

`H` should be [0 1 0 0 0 0] for Annulus measurement so that  `H * predicted_vec` becaomes [e_bound_loc1], which is phi measurement.

Comment (beomki-yeo) [reply core/include/traccc/edm/track_state.hpp:80]: No I think actually this PR is doing the right thing as the following line should bring the `e_bound_loc1` component when `subs.get_indices()[0] == 1`.

https://github.com/acts-project/traccc/blob/fd145d8fa3b5dac40fa88e449d1d235f0dff6b85/core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp#L77-L78

I guess we should be generous about using the magic number here but we can be more explicit..

```c++
if (m_measurement.subs.get_indices()[0] == 0){
            matrix_operator().element(ret, 0, 0) = m_measurement.local[0];
            if constexpr (D == 2u) {
                 matrix_operator().element(ret, 1, 0) = m_measurement.local[1];
            }
}
else if (m_measurement.subs.get_indices()[0] == 1){
            matrix_operator().element(ret, 0, 0) = m_measurement.local[1];
            if constexpr (D == 2u) {
                 matrix_operator().element(ret, 1, 0) = m_measurement.local[0];
            }
}
```

And the same goes for variances.

**Comment by shimasnd** [comment]
Codes modified according to Beomki's comments. Also modified to use `e_bound_loc0` and `e_bound_loc1` instead of 0 and 1 when checking the value of `subs.get_indices()[0]`. I guess this is more proper, if I correctly understand how subspace indices are set in other places.

**Comment by krasznaa** [comment]
I'll leave it to @beomki-yeo to give the final blessing on this one. I'll approve it once Beomki is happy with it. (I think technically I'll also need to approve it now, after requesting changes on it previously...)

File: core/include/traccc/edm/track_state.hpp
Comment (beomki-yeo) [inline core/include/traccc/edm/track_state.hpp:90]: I think we don't address such case at the moment
Let us put the warning in the if block:
```c++
assert("The measurement index out of e_bound_loc0 and e_bound_loc1 should not happen.");
```

File: core/include/traccc/edm/track_state.hpp
Comment (beomki-yeo) [inline core/include/traccc/edm/track_state.hpp:127]: Same comment

```suggestion
assert("The measurement index out of e_bound_loc0 and e_bound_loc1 should not happen.");
```

**Review by beomki-yeo — APPROVED** [review]
Looks good to me

**Review by beomki-yeo — APPROVED** [review]

**Pull request merged**
